### PR TITLE
Improve AccountSwitchGrantHandler for sub organization users

### PR DESCRIPTION
**Pull request by ChanikaRuchini**
### Proposed changes in this pull request

Improve AccountSwitchGrantHandler for sub organization users

### Related Issue

- https://github.com/wso2/product-is/issues/19310

### CODE REVIEW

File: components/org.wso2.carbon.identity.user.account.association/pom.xml [hunk 1]
```diff
@@ -88,13 +92,18 @@
...
+            <artifactId>org.wso2.carbon.identity.organization.management.service</artifactId>
+        </dependency>
     <dependency>
             <groupId>org.apache.felix</groupId>
             <artifactId>org.apache.felix.scr.ds-annotations</artifactId>
             <scope>provided</scope>
         </dependency>
     </dependencies>
 
+
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.user.account.association/pom.xml]: unnecessary new line

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java [hunk 2]
```diff
@@ -196,4 +226,46 @@ private OAuth2TokenValidationResponseDTO validateToken(String accessToken) {
         return responseDTO;
     }
 
+    private String resolveUserName (String orgId, String userId ) throws IdentityOAuth2Exception {
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]:
```suggestion
    private String resolveUserName (String orgId, String userId) throws IdentityOAuth2Exception {
```

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java [hunk 3]
```diff
@@ -72,6 +85,13 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
         }
 
         User authorizedUser = User.getUserFromUserName(validationResponseDTO.getAuthorizedUser());
+        AccessTokenDO accessTokenDO = OAuth2Util.findAccessToken(token, false);
+        if (isSubOrganizationUser(accessTokenDO)) {
+            String authorizedUseName = resolveUserName(accessTokenDO.getAuthzUser().getAccessingOrganization(),
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]:
```suggestion
            String authorizedUserName = resolveUserName(accessTokenDO.getAuthzUser().getAccessingOrganization(),
```

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java [hunk 4]
```diff
@@ -72,6 +85,13 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
...
+                    accessTokenDO.getAuthzUser().getUserName().split("@")[0]);
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: for emails as username, what's the impact of this?

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java [hunk 5]
```diff
@@ -99,9 +119,19 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
...
+        AuthenticatedUser authenticatedUser = OAuth2Util.getUserFromUserName(associatedUser.toFullQualifiedUsername());
+        if (isSubOrganizationUser(accessTokenDO)) {
+            authenticatedUser.setAccessingOrganization(accessTokenDO.getAuthzUser().getAccessingOrganization());
+            authenticatedUser.setUserResidentOrganization(accessTokenDO.getAuthzUser().getUserResidentOrganization());
+            authenticatedUser.setFederatedUser(true);
+            authenticatedUser.setFederatedIdPName(accessTokenDO.getAuthzUser().getFederatedIdPName());
+            authenticatedUser.setTenantDomain(accessTokenDO.getAuthzUser().getTenantDomain());
+            String userId = getUserId(username, tenantDomain);
+            authenticatedUser.setUserId(userId);
+            authenticatedUser.setUserName(userId + "@" + accessTokenDO.getAuthzUser().getUserResidentOrganization());
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: what's the reason to change these? what are the default values set for AuthenticatedUser in sub org ?

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java [hunk 6]
```diff
@@ -196,4 +226,46 @@ private OAuth2TokenValidationResponseDTO validateToken(String accessToken) {
...
+
+    private OrganizationManager getOrganizationManager() {
+
+        return IdentityAccountAssociationServiceDataHolder.getInstance().getOrganizationManager();
+    }
+
+    private String getUserId(String userName, String tenantDomain) throws IdentityOAuth2Exception {
+
+        try {
+
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]:
```suggestion
```

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/internal/IdentityAccountAssociationServiceComponent.java [hunk 7]
```diff
@@ -204,5 +205,23 @@ protected void setIdentityCoreInitializedEventService(IdentityCoreInitializedEve
     /* reference IdentityCoreInitializedEvent service to guarantee that this component will wait until identity core
          is started */
     }
+
```
Comment (sadilchamishka) [inline components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/internal/IdentityAccountAssociationServiceComponent.java:208]: Update the license files where the file changes are made.

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 3 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: Better to use accessTokenDO.getAuthzUser().getUserResidentOrganization()

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 4 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/252/files#diff-604290ac058e0ac40d2bd89a168f3e62a58fcbd753e58ddadf8066bfaed56701R129

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java [hunk 8]
```diff
@@ -72,6 +85,13 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
         }
 
         User authorizedUser = User.getUserFromUserName(validationResponseDTO.getAuthorizedUser());
+        AccessTokenDO accessTokenDO = OAuth2Util.findAccessToken(token, false);
+        if (isSubOrganizationUser(accessTokenDO)) {
+            String authorizedUseName = resolveUserName(accessTokenDO.getAuthzUser().getAccessingOrganization(),
+                    accessTokenDO.getAuthzUser().getUserName().split("@")[0]);
+            authorizedUser.setUserName(authorizedUseName);
+            authorizedUser.setTenantDomain(accessTokenDO.getAuthzUser().getAccessingOrganization());
```
Comment (sadilchamishka) [inline components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: Not sure changing the tenant domain is a good approach. Basically the tenant domain should be kept unchanged.

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 5 above)
Comment (sadilchamishka) [inline components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: Better if we can follow this approach https://github.com/wso2-extensions/identity-oauth2-grant-organization-switch/blob/main/components/org.wso2.carbon.identity.oauth2.grant.organizationswitch/src/main/java/org/wso2/carbon/identity/oauth2/grant/organizationswitch/OrganizationSwitchGrant.java#L102 and then set the required information like accessing organization and resident organization.
Trying to set the user-id may not be required.

Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: Any reason which persuades to set user-id. Is it due to user id handling for federated user fails for some reason. if so, thats fine as here we are treating a user who didn't federated login as the authenticated user.

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 6 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: fixed

File: components/org.wso2.carbon.identity.user.account.association/pom.xml (hunk 1 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.user.account.association/pom.xml]: Fixed

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 2 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: Fixed

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 8 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: for sub organization flow initially tenant is set as the root tenat. If this is not change validation fails.

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 5 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: yes if not changed the user id handling for federated user fails.

Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: If these not changed default values set as null. token request fails

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 3 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: Fixed

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/internal/IdentityAccountAssociationServiceComponent.java (hunk 7 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/internal/IdentityAccountAssociationServiceComponent.java:208]: Fixed

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 4 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: Fixed

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 8 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: We may have to check which validation gets failed and fixe there. Any complication on that.

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 5 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: Any comment for the first suggession.

**Review by sadilchamishka — APPROVED** [review]

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 8 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: This is sorted.

File: components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java (hunk 5 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java]: Fixed

**Pull request merged**